### ISSUE TITLE
Breaking backwards compat between v11 and v12 from differing protobuf field nums

### ISSUE THREAD

**Issue by kimlisa** [gravitational/teleport#24817]
PR https://github.com/gravitational/teleport/pull/22528 introduced a new field `MFAResponse` to protobuf message `UserCertsRequest` with struct field number `18`. When backporting I changed the field number to `17` which breaks the backwards compatibility between v12 and v11:

master - v12: https://github.com/gravitational/teleport/pull/22528/files#diff-04d701ddf274076d7060ab5a9e74cc33e438c51e953f551997715bb6bfb505fdR313

v11: https://github.com/gravitational/teleport/pull/22921/files#diff-04d701ddf274076d7060ab5a9e74cc33e438c51e953f551997715bb6bfb505fdR295

#### Reproduction (v11 proxy with v12 auth)

With per-session MFA enabled (it was tested the below works without it enabled):

ssh session file transfers:

![image](https://user-images.githubusercontent.com/43280172/233287259-3b34485b-97dd-4938-b35b-a377ce0304a8.png)

discover connection testing:

![image](https://user-images.githubusercontent.com/43280172/233287401-605c23b6-cc83-4a5c-b300-b6a3f5f6d2f0.png)

**Comment by codingllama** [comment]
Could we track adding a `buf breaking` linter to CI as part of this as well?

**Comment by espadolini** [comment]
From some tests, it seems like the breakage is limited to clusters that have per-session MFA configured, and only to SSH file transfers in the web UI with a v11 proxy and v12 auth and the step 2 of "Test Connection" in the Teleport Discover flow ("Verify that the server is accessible"); step 3 ("Connect to the server") works just fine. The per-session MFA flow through `tsh` v11 with a v12 cluster seems to be unaffected as well.

Given the very limited scope of the breakage, I'd recommend just leaving things as they are (with a comment in the v11 `authservice.proto`, lest we add more breakage in the mix) and focusing on preventing similar issues in the future.

**Comment by kimlisa** [comment]
> Could we track adding a `buf breaking` linter to CI as part of this as well?

i will look into this after I finish/test a feature for v13

**Comment by espadolini** [comment]
After doing some tests, `buf breaking` has some caveats:
- it doesn't detect breaking changes to custom options, which for us means no breaking change detection around `gogoproto.jsontag` (it's something that they're working on, but it's unclear if it's going to be supporting arbitrary options or just specific options like `google.api.http`);
- we have the terrible habit of renaming message types (v11 has `types.RoleV5` and v12+ have `types.RoleV6`, v12 had `proto.WatchKind` but we moved it to `types.WatchKind` for v13+) even when there's no reason for it, and `buf breaking` doesn't really handle that.

Ignoring the `gogoproto.jsontag` issue (which never resulted in a breakage so far), we can work around the break in `WatchKind` and in `UserCertsRequest` (the object of this issue) by awkwardly moving them to a separate file and ignoring the specific breaking changes in it, and we can avoid the issue about renaming messages by _not renaming messages_. We can rename `RoleV5` to `RoleV6` in `branch/v11` and add a type alias on the go side, or maybe add an ignore for `RoleV5` (which would need a separate `.proto`).

With that (and some cleanups here and there) we should be able to run `buf breaking` against the base branch of the PR for all PRs, and on `master` against the PR (note the order) for PRs to release branches.

**Comment by espadolini** [comment]
Another caveat: with `WIRE_JSON` or `WIRE` breaking detection, we will also not be warned about the deletion of RPCs. We could add the `RPC_NO_DELETE` rule to the list, but we actually delete RPCs occasionally (after replacing them, and only after a couple of versions), and that'd prevent us from doing so in the future.

**Comment by codingllama** [comment]
Tl;DR: +1 for `buf breaking`, it sounds like we have some reasonable ideas of how to deal with linter issues.

> it doesn't detect breaking changes to custom options (...)

That is alright, it doesn't leave us any worse than before.

> we have the terrible habit of renaming message types (...)

Renames _are_ breaking, and typically poor-form in the proto world, where most changes are meant to be backwards compatible. It's a peculiarity of how Teleport handles versioning, so it was bound to be flagged.

Having separate proto files seems like a win to me anyway, and we could use those to have more fine-grained control over linters, like you said. Not renaming messages would be my preference too, but it seems unrealistic to change the way Teleport manages protos this suddenly.

> (...) we will also not be warned about the deletion of RPCs (...)

Also doesn't leave us any worse than before, so seems OK. Ideally we'd turn on the linter so we can flag "accidental" deletion, but we'll likely want a way to push our cleanups forward.

**Comment by zmb3** [comment]
Closing this, as `buf breaking` support is being tracked in #26688
